oslabs-beta/watchmo
# server serves parsed data for the frontend

Significant adjustments have been made to the structure of the command line tools, so that the data can be easily accessed on the front-end. 

**What's changed?**
1. No more init! Instead, the timing data is stored within the package itself in a built in directory, '/src/watchmoData', which contains the files 'snapshots.txt', 'config.json', and once parsed, 'parsedData.json'.

2. The watch command now writes the data in a different format and with a different structure.
```
// new structure for the saved data in snapshots.js, each individual query/response
// This is NOT what the front end will receive, 
// this is what is passed between watch and mo via snapshots.txt.
{
  "category": "[some string]",  // The name of the query's associated category
  "data" : {
    "timestamp": new Date(), //gives a timestamp for the query
    "response": {"[some GQL data]"},
    "query" : "[some query string]",
    "timing" : "[Array containing timing information]"
    }
}
```
 The raw data is stored as a single .txt file (_not_ JSON) where every additional query/response is appended to the end of this file. These query/responses are demarcated by a special string denoted by the variable `DEMARCATION` in watch.js. Currently this string is * W * M * O *, for fun. 

3. Mo does mo'. In particular, mo parses the raw data in snapshots.txt, bundles it into a JSON file 'parsedData.json' with a more useful shape, and opens up the webpage.
The structure of this JSON file is as follows: 
```
//This WILL be sent to the frontend. It can be accessed via the http route '/data'
{
  "[first category name]" : {
    "[first query for this category]" : [ {
      "timestamp": "[string containing the time the query was sent]",
      "query" : "[the query string sent to the server]",
      "response" : "[the actual GQL response]",
      "timing" : "[the time it took to get the GQL response]"  
    }]
  }
  //same for each other category...
 }
```

4. The server now serves two new routes, '/config' and '/data', which send the config.json and parsedData.json files.

The motivation behind these changes is twofold:
- The server needs to serve a json file with a structure that is easy for the frontend to work with.
- Storing timing data has to go fast, otherwise the Intervals may overlap and errors will occur. Appending data to a textfile is much faster than parsing a JSON file, adding to it, and writing that whole file again. 

**How to test it** 
_Setup_
1. Adjust your config.json file. Make sure the endpoint is correct, and the queries you want are correct.
2. If you haven't yet, run `npm link` in the watchmo root directory. 
3. Make sure that the endpoint server is running. If you are using Evan's GQL test server, navigate to that directory and run `npm start`.
4. Make sure that you have the file "src/watchmoData", and that there is only a config.json file in this directory. If there are snapshot directories, parsedData, or anything else, delete it. 

_Using watchmo_
1. Run `watchmo watch` in the command line and wait. You should see 'file saved in [path]' in the terminal after some time.
2. Check the snapshots.txt file in "src/watchmoData". It should be a mess of text with * W * M * O * interspersed between object literals.
3. Run `watchmo mo` in the command line. You should see 'DATA BUNDLED' in the terminal, and your browser should open.
4. Check the parsedData.json file in "src/watchmoData". You should see a large JSON object.

_Checking the Routes_
1. Run `npm start` in the terminal. You should see "app listening on port [some number]". 
2. Use Postman or some similar tool to check the route "http://localhost:[port]/config". You should get the config.json file back.
3. Use Postman or some similar tool to check the route "http://localhost:[post]/data". You should get the parsedData.json file back. 


**Notes**
- The config file has been populated with faux settings for easy testing. DO NOT LET THIS BE UPLOADED IN THE FINAL PRODUCTION DRAFT. 
- the .gitignore file has been updated to include the new data files. These also should not be pushed during production.
